src/ground_truth_nsrts.py

```diff
@@ -853,36 +946,127 @@ def placeinshelf_sampler(state: State, goal: Set[GroundAtom],
+            preconditions = {
+                LiftedAtom(Holding, [obj]),
+                LiftedAtom(holding_rot_pred, [obj])
+            }
+            add_effects = {
+                LiftedAtom(GripperOpen, [robot]),
+                LiftedAtom(OnTable, [obj])
+            }
+
         delete_effects = {
             LiftedAtom(Holding, [obj]),
             LiftedAtom(holding_rot_pred, [obj]),
         }
+        if CFG.env == "painting":
+            side_predicates = set()
+        elif CFG.env == "repeated_nextto_painting":
+            side_predicates = {NextTo, NextToNothing}
 
         def placeontable_sampler(state: State, goal: Set[GroundAtom],
                                  rng: np.random.Generator,
                                  objs: Sequence[Object]) -> Array:
-            # Always release the object where it is, to avoid the possibility
-            # of collisions with other objects.
-            del goal, rng  # unused
+            del goal  # unused
             x = state.get(objs[0], "pose_x")
-            y = state.get(objs[0], "pose_y")
-            z = state.get(objs[0], "pose_z")
+            if CFG.env == "painting":
+                # Always release the object where it is, to avoid the
+                # possibility of collisions with other objects.
+                y = state.get(objs[0], "pose_y")
+                z = state.get(objs[0], "pose_z")
+            elif CFG.env == "repeated_nextto_painting":
+                # Release the object at a randomly-chosen position on the table
+                # such that it is NextTo the robot.
+                robot_y = state.get(objs[1], "pose_y")
+                y_sample_lb = max([
+                    RepeatedNextToPaintingEnv.table_lb,
+                    robot_y - RepeatedNextToPaintingEnv.nextto_thresh
+                ])
+                y_sample_ub = min([
+                    RepeatedNextToPaintingEnv.table_ub,
+                    robot_y + RepeatedNextToPaintingEnv.nextto_thresh
+                ])
+                y = rng.uniform(y_sample_lb, y_sample_ub)
+                z = RepeatedNextToPaintingEnv.obj_z
             return np.array([x, y, z], dtype=np.float32)
 
         placeontable_nsrt = NSRT(f"PlaceOnTable-{holding_rot_pred.name}",
                                  parameters, preconditions, add_effects,
-                                 delete_effects, set(), option, option_vars,
-                                 placeontable_sampler)
+                                 delete_effects, side_predicates, option,
+                                 option_vars, placeontable_sampler)
         nsrts.add(placeontable_nsrt)
 
+    if CFG.env == "repeated_nextto_painting":
+
+        def moveto_sampler(state: State, goal: Set[GroundAtom],
+                           _rng: np.random.Generator,
+                           objs: Sequence[Object]) -> Array:
+            del goal  # unused
+            y = state.get(objs[1], "pose_y")  # + rng.uniform() * 0.4
```

Comment:
what's going on with the commented out thing here? let's either remove or add a comment describing when we might want to re-enable it in the future?

Reply:
oh good question; i think @wmcclinton left that in and forgot to delete; will remove!